### DISCUSSION TITLE
Passing ref as props on children Element

### DISCUSSION THREAD

**Discussion by jodespanola** [vercel/next.js#64298] [Help] [answered]
### Summary

Hi, I have these two component- one is the parent element and the other a button component. My goal is to pass in the ref of an element to the child component and use that do an animation when the button component is clicked. But the problem that I'm encountering is that the ref seems to be null every time. I think its maybe because the component has not been mounted yet, but I don't know how to make the child component to wait for it to be mounted.
```
// parent component

"use client"

import TransitionLayoutfrom "@/components/transitionLayout";
import TransitionLink from "@/components/transitionLink"; 
import { useContext, useEffect, useRef } from "react";

export default function Home() {
  const transition_div = useRef<HTMLDivElement | null>(null);

  return (
    <TransitionLayout>
    <div
        ref={transition_div} 
        className="min-h-screen w-screen bg-rose-950 absolute -left-full">
    </div>
      <main className="min-h-screen flex flex-col justify-center items-center gap-4">
        <h1 className="text-lg">
          About
        </h1>
        <div className="flex justify-center gap-10 items-center">
          <TransitionLink transition_div={transition_div.current} href='./home' label='Home'/>
          <TransitionLink transition_div={transition_div.current} href='./service' label='Service'/>
        </div>
      </main>
    </TransitionLayout>
  );
}
```

```
// transitionLink.tsx

"use client"

import { animatePageOut } from "@/utils/animation";
import { useRouter } from "next/navigation"

interface TransitionLinkProps {
    href: string,
    label: string,
    transition_div: HTMLDivElement | null,
}

export default function TransitionLink({
    href,
    label,
    transition_div
}: TransitionLinkProps ) {
    const router = useRouter();

    let handleClick = () => {
        console.log(transition_div)
        animatePageOut(transition_div, href, router);
    }

    return (
        <button
            className="cursor-pointer"
            onClick={handleClick}
        >
            {label}
        </button>
    )
}
```

**Comment by icyJoseph** [comment, marked as answer]
There is a bunch of issues here, first, you are copying here:

```jsx
 <TransitionLink transition_div={transition_div.current} href='./home' label='Home'/>
```

So `transition_div` is equal to `null`, not the transition ref object anymore. Since refs are stable objects across renderings, and changes to refs, don't trigger re-renders, there's no way to pass the updated ref downstream. 

It is like doing this:

```js
const ref = { current: 2 }
const value = ref.current; // here we copy the current into value
ref.current = 100; // current is updated elsewhere
console.log(value !== ref.current); // true 
```

A simple fix would be:

```tsx
 <TransitionLink transition_div={transition_div} href='./home' label='Home'/>
```

And of course, update your types:

```tsx
interface TransitionLinkProps {
  href: string;
  label: string;
  transition_div: RefObject<HTMLDivElement | null>;
}
```

And while that might work, I'd say it is not optimal. It is a convention among React devs to pass refs, as a `ref` prop, but the `ref` prop is not passed, up to React 18, by default, you have to forward it:

```tsx
'use client';

import { forwardRef } from 'react';

interface TransitionLinkProps {
  href: string;
  label: string;
}

const TransitionLink = forwardRef<HTMLDivElement, TransitionLinkProps>(
  function TransitionLink({ label }, ref) {
    let handleClick = () => {
      console.log(ref);
    };

    return (
      <button className="cursor-pointer" onClick={handleClick}>
        {label}
      </button>
    );
  }
);
```

This way, you can invoke `TransitionLink` with a ref prop:

```tsx
'use client';

import { useRef, ElementRef } from 'react';
import { TransitionLink } from 'somewhere';

export default function Home() {
  const transition_div = useRef<ElementRef<typeof TransitionLink>>(null);

  return (
    <section>
      <div
        ref={transition_div}
        className="min-h-screen w-screen bg-rose-950 absolute -left-full"
      ></div>
      <main className="min-h-screen flex flex-col justify-center items-center gap-4">
        <h1 className="text-lg">About</h1>
        <div className="flex justify-center gap-10 items-center">
          <TransitionLink ref={transition_div} href="./home" label="Home" />
        </div>
      </main>
    </section>
  );
}

```

And that'll make it work nicely. Since the refs is being accessed within a callback, you'll have it by the time the callback is called.

The key insight is that a `ref` is an object that is constant across re-renders, and it has a `current` property attached to it, where the value is stored. If you copy the value into a variable, and pass that along, that variable won't be updated when `current` is changed to something else in the original `ref`, excluding mutations from this formulation.

Here's a playground with the code above: https://stackblitz.com/edit/nextjs-hxv9t7?file=app%2Fpage.tsx

Please do read these links, to get a deeper understanding of refs:

- https://react.dev/learn/referencing-values-with-refs
- https://react.dev/reference/react/useRef

**Reply by jodespanola** [reply]
Hi, Thanks for the answer and the references. I was looking at forward refs earlier but seems like I had a entirely different understanding of its use. I'll read to those references, thanks again for the help!
